### PR TITLE
Implement TLS bootstrap for kubelet using `--experimental-bootstrap-kubeconfig`  (2nd take)

### PR DESCRIPTION
**Pull request by yifan-gu**
Ref kubernetes/features#43 (comment)

cc @gtank @philips @mikedanese @aaronlevy @liggitt @deads2k @errordeveloper @justinsb 

Continue on the older PR https://github.com/kubernetes/kubernetes/pull/30094 as there are too many comments on that one and it's not loadable now.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30922)

### CODE REVIEW

**Comment by mikedanese** [comment]
> there are too many comments on that one and it's not loadable now.

:) we should switch to gerrit

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -335,6 +335,17 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
 
 	if kcfg == nil {
 		var kubeClient, eventClient *clientset.Clientset
+
+		if s.KubeConfig.Value() != "" && s.BootstrapKubeconfig != "" {
```
Comment (yifan-gu) [inline cmd/kubelet/app/server.go]: @liggitt IIUC `KubeConfig.Value()` is always non-empty

**Comment by yifan-gu** [comment]
So I pulled from @liggitt's branch for the refactoring work (thank you @liggitt !!), and addressed some additional comments. The only outstanding one left is:

https://github.com/kubernetes/kubernetes/pull/30094#discussion_r75248955:

> liggitt added a note 16 hours ago
not sure about this... do all cloud providers tolerate being initialized twice?

Is the `cloudprovider.InitCloudProvider()` idempotent?

**Comment by errordeveloper** [comment]
cc @lukemarsden

**Comment by yifan-gu** [comment]
@liggitt I am going to move the cloud initialization to before bootstrap process.

**Comment by liggitt** [comment]
"I am okay with these commits being contributed to Google" or you can squash my commit in

Moving the real cloud provider init earlier is probably better, I just hadn't looked at everything that happened in there to be sure

File: cmd/kubelet/app/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+		return nil, false, fmt.Errorf("error writing key to %s: %v", keyPath, err)
+	}
+	return generatedData, true, nil
+}
+
+// requestClientCertificate will create a certificate signing request and send it to API server,
+// then it will watch the object's status, once approved by API server, it will return the API
+// server's issued certificate (pem-encoded). If there is any errors, or the watch timeouts,
+// it will return an error.
+func requestClientCertificate(client unversionedcertificates.CertificateSigningRequestInterface, privateKeyData []byte, nodeName string) (certData []byte, err error) {
```
Comment (errordeveloper) [inline cmd/kubelet/app/bootstrap.go]: Any chance you could make this one public or it's unsuitable as is?

Comment (yifan-gu) [inline cmd/kubelet/app/bootstrap.go]: @errordeveloper Open to make it public, but not sure who else will be using it?

Comment (errordeveloper) [inline cmd/kubelet/app/bootstrap.go]: We'd like to make CSRs from kubeadm.

Comment (yifan-gu) [inline cmd/kubelet/app/bootstrap.go]: @errordeveloper ACK, fixed.

**Comment by mikedanese** [comment]
@gtank can you say "I am okay with these commits being contributed to Google"

**Comment by gtank** [comment]
I am okay with these commits being contributed to Google

File: cmd/kubelet/app/bootstrap.go (hunk 2 above)
Comment (errordeveloper) [inline cmd/kubelet/app/bootstrap.go]: Thank you!

Comment (errordeveloper) [inline cmd/kubelet/app/bootstrap.go]: @lukemarsden FYI.

**Comment by mikedanese** [comment]
@calebamiles

> How are you feeling about things @mikedanese, are we fine with a merge once our test failures are resolved?

I'm ok with merging this if/once this feature doesn't affect currently flows (e.g. users not using TLS bootstrap) and it actual (mostly) works. CI also needs to be green.

> In a similar vein, @smarterclayton could we address your desired UX in a separate feature. If we're imagining that this would be an experimental feature do you think it could make sense to let some users play around with the proposed workflow and address further UX simplification of cluster bootstrapping with work like #30360 or in a discussion originating out of sig-cluster-lifecycle if that's the appropriate place. I'm also wondering, @detiber if your desires could be addressed as an iteration on some future work to enable automated approval.

I agree with this sentiment. We are 90% to alpha for this feature. More discussion and revision at this point (which likely means holding this feature for another release) is going to be a net negative in the long run as we will lose out on a release of field testing. We've strapped alpha and experimental on this feature. Our hands aren't tied to incrementally improve this feature or rip it out altogether.

File: cmd/kubelet/app/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, fmt.Errorf("cannot create certificate signing request: %v", err)
+
+	}
+
+	// Make a default timeout = 3600s.
+	var defaultTimeoutSeconds int64 = 3600
+	resultCh, err := client.Watch(api.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &defaultTimeoutSeconds,
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", req.Name),
```
Comment (yifan-gu) [inline cmd/kubelet/app/bootstrap.go]: With this being set, I got errors 
```
Field selector: certificates/v1alpha1 - certificatesigningrequests - metadata.name - csr-dxmh8: need to check if this is versioned correctly.
Error: failed to run Kubelet: cannot watch on the certificate signing request: No field label conversion function found for version: certificates/v1alpha1
failed to run Kubelet: cannot watch on the certificate signing request: No field label conversion function found for version: certificates/v1alpha1
```

Not sure what's going on here? @liggitt

**Comment by yifan-gu** [comment]
Manually tested (steps: https://gist.github.com/yifan-gu/b62588dab290bbb2f1d91e8791ae21d4 ), it works with one questions I posted above.

File: cmd/kubelet/app/bootstrap.go (hunk 3 above)
Comment (mikedanese) [inline cmd/kubelet/app/bootstrap.go]: Predicates for field selectors need to be written and registerd manually in the apiserver e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/pod/strategy.go#L193

ref https://github.com/kubernetes/kubernetes/issues/1362 https://github.com/kubernetes/kubernetes/pull/28112

Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: fixed in https://github.com/kubernetes/kubernetes/pull/30950

File: cmd/kubelet/app/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate certificate request: %v", err)
+	}
+
+	req, err := client.Create(&certificates.CertificateSigningRequest{
+		TypeMeta:   unversioned.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: api.ObjectMeta{GenerateName: "csr-"},
+
+		// Username, UID, Groups will be injected by API server.
+		Spec: certificates.CertificateSigningRequestSpec{Request: csr},
```
Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: TODO here for indicating in the `Spec` that this is a request for a cert with allowed usage of "client auth" (in cfssl parlance)
@mikedanese @gtank, any update on how you'd like to express that in the CertificateSigningRequestSpec?

File: cmd/kubelet/app/server.go [hunk 5]
```diff
@@ -335,6 +335,33 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
...
+			}
+			if cloud == nil {
+				glog.V(2).Infof("No cloud provider specified: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
+			} else {
+				glog.V(2).Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
+			}
+		}
+
+		if s.BootstrapKubeconfig != "" {
+			nodeName, err := getNodeName(kcfg.Cloud, nodeutil.GetHostname(s.HostnameOverride))
```
Comment (liggitt) [inline cmd/kubelet/app/server.go]: won't kcfg be nil here?

File: cmd/kubelet/app/server.go [hunk 6]
```diff
@@ -439,13 +454,26 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
 	return nil
 }
 
+// getNodeName returns the node name according to the cloud provider
+// if cloud provider is specified. Otherwise, returns the hostname of the node.
+func getNodeName(cloud cloudprovider.Interface, hostname string) (string, error) {
```
Comment (liggitt) [inline cmd/kubelet/app/server.go]: having this factored out is good, but we should use it in RunKubelet as well or we'll drift from how the node name is computed for real. pushed a commit to https://github.com/liggitt/kubernetes/commits/tls-bootstrap that unifies those two paths (and fixes the NPE at https://github.com/kubernetes/kubernetes/pull/30922/files/4120179db5ce6c3fadc872a268347f5be038f98f#r75416356)

File: cmd/kubelet/app/bootstrap.go (hunk 4 above)
Comment (mikedanese) [inline cmd/kubelet/app/bootstrap.go]: Until we have a concept of signing profiles, can we only allow client certs? It's seems too late to design this.

File: pkg/util/certificates/csr.go [hunk 7]
```diff
@@ -38,3 +45,95 @@ func ParseCertificateRequestObject(obj *certificates.CertificateSigningRequest)
...
+		}
+	}
+}
+
+// NewCertificateRequest generates a PEM-encoded CSR using the supplied private key, subject, and SANs.
+// privateKey must be a *ecdsa.PrivateKey or *rsa.PrivateKey.
+func NewCertificateRequest(privateKey interface{}, subject *pkix.Name, dnsSANs []string, ipSANs []net.IP) (csr []byte, err error) {
+	var sigType x509.SignatureAlgorithm
+
+	switch privateKey := privateKey.(type) {
```
Comment (liggitt) [inline pkg/util/certificates/csr.go]: @gtank is there a reason the [default signature CreateCertificateRequest computes](https://golang.org/src/crypto/x509/x509.go?s=47171:48866#L1492) from the passed-in key wasn't working well?

File: pkg/util/certificates/csr.go [hunk 8]
```diff
@@ -38,3 +45,95 @@ func ParseCertificateRequestObject(obj *certificates.CertificateSigningRequest)
...
+	case *ecdsa.PrivateKey:
+		switch privateKey.Curve.Params().BitSize {
+		case 512:
```
Comment (liggitt) [inline pkg/util/certificates/csr.go]: @gtank this was `case 521:` in the original PR, which I assumed was a typo. this is switching on bit size, not on the curve (for which 521 would be [valid](https://golang.org/pkg/crypto/elliptic/#P521)). can you confirm 512 is the right case here (assuming we still need to be computing the signature algorithm ourselves)

File: cmd/kubelet/app/bootstrap.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+			if !success {
+				if err := os.Remove(keyPath); err != nil {
+					glog.Warningf("Cannot clean up the key file %q: %v", keyPath, err)
+				}
+			}
+		}()
+	}
+
+	// Get the cert.
+	certPath, err := filepath.Abs(filepath.Join(certDir, "kubelet-client.crt"))
```
Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: nit: I forgot to use the defaultKubeletClientCertificateFile constant defined above (same for key)

File: cmd/kubelet/app/bootstrap.go (hunk 4 above)
Comment (mikedanese) [inline cmd/kubelet/app/bootstrap.go]: > can we only allow client cert allowed use?

Client auth and signing which is useful during the approval challenge.

Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: I suppose that's fine... I'm sort of counting on the alpha-ness of the current API if we end up deciding the spec.usage field should be required

Comment (mikedanese) [inline cmd/kubelet/app/bootstrap.go]: If we are going to surface raw allowed usage in the API, I would see why we shouldn't just use the x509 extension. Would we ever sign a CSR without the allowed usages that it requested? Would we ever sign a CSR with allowed usages that it didn't request?

I was imaging that the "signing profile" would be closer to what we do with "storage-class" for PDs. For storage, you have silver, platinum, gold... For certs, you would have signature, client, server... The benefit is that configuration matrix (which is sparse w.r.t. valid configurations) is hidden from the user.  If we imagine that the only thing we'll ever tweak in a storage class is allowed usage, then signing profiles are not useful and we should expose allowed usage directly.

I think this conversation is very important but I don't think there is harm in deferring it. The development of the "experimental" api was motivated by allowing this type of experimentation so let's take advantage of the alpha-ness.

Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: > Would we ever sign a CSR without the allowed usages that it requested?

Probably not

> Would we ever sign a CSR with allowed usages that it didn't request?

Maybe (e.g. add in "signing" even if just client cert was requested)? not sure

> I think this conversation is very important but I don't think there is harm in deferring it.

sure. given we can scope auto-approval to node requests for a cert of a certain shape, and that the kubelet isn't using the obtained cert for double-duty, I think I'd be ok leaving the default profile as-is. when we figure out how to express requested usage, we can tighten the auto-approval and the kubelet request and nothing else has to change.

File: cmd/kubelet/app/bootstrap.go (hunk 3 above)
Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: you can wait for https://github.com/kubernetes/kubernetes/pull/30950 or comment out `FieldSelector:  fields.OneTermEqualSelector("metadata.name", req.Name),` with a TODO to re-enable (it'll be an inefficient watch, but the uid check will limit us to processing our own CSR)

File: cmd/kubelet/app/bootstrap.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	return generatedData, true, nil
+}
+
+// RequestClientCertificate will create a certificate signing request and send it to API server,
+// then it will watch the object's status, once approved by API server, it will return the API
+// server's issued certificate (pem-encoded). If there is any errors, or the watch timeouts,
+// it will return an error.
+func RequestClientCertificate(client unversionedcertificates.CertificateSigningRequestInterface, privateKeyData []byte, nodeName string) (certData []byte, err error) {
+	subject := &pkix.Name{
+		Organization: []string{"system:nodes"},
```
Comment (mikedanese) [inline cmd/kubelet/app/bootstrap.go]: @liggitt remind me again, we use OrganizationalUnit for groups, CommonName for user. What is Organization used for?

Comment (liggitt) [inline cmd/kubelet/app/bootstrap.go]: https://github.com/kubernetes/kubernetes/pull/30392 pulls from Organization

File: cmd/kubelet/app/bootstrap.go (hunk 3 above)
Comment (yifan-gu) [inline cmd/kubelet/app/bootstrap.go]: @liggitt ACK

File: cmd/kubelet/app/server.go (hunk 5 above)
Comment (yifan-gu) [inline cmd/kubelet/app/server.go]: Hmm, thought I fixed this, maybe forgot to push.

File: cmd/kubelet/app/server.go (hunk 6 above)
Comment (yifan-gu) [inline cmd/kubelet/app/server.go]: Fixed.

File: cmd/kubelet/app/bootstrap.go (hunk 9 above)
Comment (yifan-gu) [inline cmd/kubelet/app/bootstrap.go]: fixed

File: cmd/kubelet/app/bootstrap.go (hunk 4 above)
Comment (gtank) [inline cmd/kubelet/app/bootstrap.go]: Yeah, I'm still experimenting with this. Go's CertificateRequest support only knows about [pkix.Extension](https://godoc.org/crypto/x509/pkix#Extension) types, which are pretty raw and- as far as I can tell- under-specified for expressing desired key usages in PKCS#10. 

I agree that requested usage is a good thing to expose, but we shouldn't block this feature on something that might need upstream Go changes. I'll take a closer look at what openssl does when I get a chance.

File: pkg/util/certificates/csr.go (hunk 7 above)
Comment (gtank) [inline pkg/util/certificates/csr.go]: It defaults all RSA to SHA256WithRSA, which isn't what we want as key sizes increase.

File: pkg/util/certificates/csr.go (hunk 8 above)
Comment (gtank) [inline pkg/util/certificates/csr.go]: No, this is catching [NIST P-521](https://golang.org/src/crypto/elliptic/elliptic.go#L358) which is indeed a 521-bit curve. `512` is an unfortunate and confusing cognate.

Comment (gtank) [inline pkg/util/certificates/csr.go]: Admittedly, this code could make those numbers more obviously intentional like [cfssl](https://github.com/cloudflare/cfssl/blob/master/csr/csr.go#L102) and the [upstream function](https://golang.org/src/crypto/x509/x509.go?s=47171:48866#L1492) you pointed out do.

Comment (liggitt) [inline pkg/util/certificates/csr.go]: ah, sorry for the mixup. +1 to following upstream for the ecdsa case:
```
switch privateKey.Curve {
case elliptic.P224(), elliptic.P256():
  sigType = x509.ECDSAWithSHA256
case elliptic.P384():
  sigType = x509.ECDSAWithSHA384
case elliptic.P521():
  sigType = x509.ECDSAWithSHA512
default:
  err = errors.New("x509: unknown elliptic curve")
}
```

Comment (liggitt) [inline pkg/util/certificates/csr.go]: or just customizing the sigType in RSA cases, and letting the upstream code handle ECDSA cases itself

**Comment by liggitt** [comment]
changes LGTM

**Comment by yifan-gu** [comment]
@mikedanese How to get the cla bot green, should @liggitt @gtank reply the bot?

**Comment by liggitt** [comment]
the `cla:human-approved` label will satisfy the merge bot

**Comment by yifan-gu** [comment]
Tested manually after squashing and pulling #30950. Works fine :)

**Pull request merged**